### PR TITLE
Allow node removal upon role change from worker to master

### PR DESCRIPTION
**Pull request by shreesha21**
When the role of the node changes from worker to non-worker, the manager pod will be stopped and as a result node removal was resulting in failure. The following code changes support removing node in such situations.

IssueURL: https://github.com/rancher/longhorn/issues/453

### CODE REVIEW

File: controller/node_controller.go [hunk 1]
```diff
@@ -381,6 +381,7 @@ func (nc *NodeController) syncNode(key string) (err error) {
 		condition.Reason = string(types.NodeConditionReasonManagerPodMissing)
 		condition.Message = fmt.Sprintf("manager pod missing: node %v has no manager pod running on it", node.Name)
 		node.Status.Conditions[types.NodeConditionTypeReady] = condition
+		node.Spec.AllowScheduling = false
```
Comment (yasker) [inline controller/node_controller.go]: User can still set scheduling to false when the node is offline. So we shouldn't need to do this.

File: manager/node.go [hunk 2]
```diff
@@ -128,8 +128,9 @@ func (m *VolumeManager) DeleteNode(name string) error {
 	}
 	condition := types.GetNodeConditionFromStatus(node.Status, types.NodeConditionTypeReady)
 	// Only could delete node from longhorn if kubernetes node missing
-	if condition.Status == types.ConditionStatusTrue || condition.Reason != types.NodeConditionReasonKubernetesNodeDown ||
-		node.Spec.AllowScheduling || len(replicas) > 0 || len(engines) > 0 {
+	if condition.Status == types.ConditionStatusTrue || (condition.Reason != types.NodeConditionReasonKubernetesNodeDown &&
```
Comment (yasker) [inline manager/node.go]: ManagerPodMissing is different from it has been converted to non-worker. It can be just temporarily, e.g during upgrade or just started up.

Btw, we should rename `KubernetesNodeDown` to `KubernetesNodeGone`.

And previous the condition for `NodeGone` is only we cannot find it. But if we found it's nonworker, we should also consider it gone.

Comment (shreesha21) [reply manager/node.go]: How can we identify if the node is non-worker? 
I was curious to know where do we maintain the role of the node? Could you please point me to the right attribute?

Comment (yasker) [reply manager/node.go]: It's kubernetes node. If you see `k get node` you will see the Roles.

This is Kubernetes node in the Longhorn node controller: https://github.com/rancher/longhorn-manager/blob/master/controller/node_controller.go#L387

Comment (yasker) [reply manager/node.go]: You would need to check Kubernetes API to find out where is the role stored and use that value for deciding if we should declare the node is gone (in additional to the node is really gone)

Comment (yasker) [reply manager/node.go]: An easy way is looking at `kubectl get node -o yaml`.

File: controller/kubernetes_controller_test.go [hunk 3]
```diff
@@ -558,7 +558,7 @@ func (s *TestSuite) TestDisasterRecovery(c *C) {
 	tc = generateDisasterRecoveryTestCaseTemplate()
 	tc.volume.Status.State = types.VolumeStateDetached
 	tc.pvc.Status.Phase = apiv1.ClaimBound
-	tc.node = newNode(TestNode1, TestNamespace, true, types.ConditionStatusFalse, types.NodeConditionReasonKubernetesNodeDown)
+	tc.node = newNode(TestNode1, TestNamespace, true, types.ConditionStatusFalse, types.NodeConditionReasonKubernetesNodeGone)
```
Comment (yasker) [inline controller/kubernetes_controller_test.go]: Put the rename change into a separate commit.

File: controller/node_controller.go [hunk 4]
```diff
@@ -403,6 +407,32 @@ func (nc *NodeController) syncNode(key string) (err error) {
 			return err
 		}
 	} else {
+		// check if the role of the node got changed
+		kubeNodeLabels := kubeNode.GetLabels()
+		if len(kubeNodeLabels) != 0 {
+			isCurrentNodeWorker := false
```
Comment (yasker) [inline controller/node_controller.go]: This is not `CurrentNode`. Node controller check for all nodes since if the node is lost, the manager won't be running.

File: controller/node_controller.go [hunk 5]
```diff
@@ -403,6 +407,32 @@ func (nc *NodeController) syncNode(key string) (err error) {
...
+
+			for label, value := range kubeNodeLabels {
```
Comment (yasker) [inline controller/node_controller.go]: GetLabels() returns a map. Just get the value directly.

File: controller/node_controller.go [hunk 6]
```diff
@@ -403,6 +407,32 @@ func (nc *NodeController) syncNode(key string) (err error) {
...
+				if strings.Compare(label, nodeRoleWorkerLabel) == 0 && strings.Compare(value, "true") == 0 {
```
Comment (yasker) [inline controller/node_controller.go]: Use strconv.ParseBool

File: controller/node_controller.go [hunk 7]
```diff
@@ -403,6 +407,32 @@ func (nc *NodeController) syncNode(key string) (err error) {
...
+					break
+				}
+			}
+
+			if !isCurrentNodeWorker {
+				// role of the node got changed to non-worker
+				condition := types.GetNodeConditionFromStatus(node.Status, types.NodeConditionTypeReady)
+				if condition.Status != types.ConditionStatusFalse {
+					condition.LastTransitionTime = util.Now()
+					nc.eventRecorder.Eventf(node, v1.EventTypeWarning, types.NodeConditionReasonKubernetesNodeGone, "Kubernetes node role changed: node %v has been moved from worker to non-worker", node.Name)
```
Comment (yasker) [inline controller/node_controller.go]: `Kubernetes node role changed: node %v is no longer a worker`

File: controller/node_controller.go [hunk 8]
```diff
@@ -403,6 +407,25 @@ func (nc *NodeController) syncNode(key string) (err error) {
 			return err
 		}
 	} else {
+		// check if the role of the node got changed to non-worker
+		kubeNodeLabels := kubeNode.GetLabels()
+		if len(kubeNodeLabels) != 0 {
```
Comment (yasker) [inline controller/node_controller.go]: If there is no label, we would have a problem too?

Also, what if the label exists but the value is false?

Comment (shreesha21) [reply controller/node_controller.go]: Yes. If we don't have a label, then integration test is failing as the test nodes that they create wouldn't have label in them.

I haven't seen any instance in my testing where the label is present and the value is false. Is this a possibility too?

Comment (yasker) [reply controller/node_controller.go]: I think if we cannot find labels, we should print out an error at least. It's very unlikely to happen but we don't want to fail silently. 

About the false value, it's possible since it's totally implementation dependent. It's better if we check for it too.

File: controller/node_controller.go [hunk 9]
```diff
@@ -403,6 +407,27 @@ func (nc *NodeController) syncNode(key string) (err error) {
...
+				}
+				condition.Status = types.ConditionStatusFalse
+				condition.Reason = string(types.NodeConditionReasonKubernetesNodeGone)
+				condition.Message = fmt.Sprintf("Kubernetes node role changed: node %v is no longer a worker", node.Name)
+				node.Status.Conditions[types.NodeConditionTypeReady] = condition
+				// set node unschedulable
+				node.Spec.AllowScheduling = false
+			}
+		} else {
+			logrus.Errorf("Error: Kubernetes node does not have any labels.")
```
Comment (yasker) [inline controller/node_controller.go]: Which node?

File: controller/node_controller.go [hunk 10]
```diff
@@ -403,6 +407,27 @@ func (nc *NodeController) syncNode(key string) (err error) {
 			return err
 		}
 	} else {
+		// check if the role of the node got changed to non-worker
+		kubeNodeLabels := kubeNode.GetLabels()
+		if len(kubeNodeLabels) != 0 {
+			if value, labelPresent := kubeNodeLabels[nodeRoleWorkerLabel]; labelPresent == false || value == "false" {
```
Comment (yasker) [inline controller/node_controller.go]: Use `strconv.ParseBool` for value, otherwise you can miss e.g. `False`, `FALSE`
